lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -0,0 +1,1499 @@
+            try:
+                certificate_object = x509.load_pem_x509_certificate(
+                    data=provider_certificate.certificate.encode()
+                )
+            except ValueError as e:
+                logger.error("Could not load certificate - Skipping: %s", e)
+                continue
+            validity_start_time = certificate_object.not_valid_before_utc
+            expiry_notification_time = calculate_expiry_notification_time(
+                validity_start_time=validity_start_time,
+                expiry_time=certificate_object.not_valid_after_utc,
+                provider_recommended_notification_time=provider_certificate.recommended_expiry_notification_time,
+            )
+            revoked = provider_certificate.revoked
+            provider_certificate = ProviderCertificate(
+                relation_id=relation.id,
+                application_name=relation.app.name,
+                csr=provider_certificate.certificate_signing_request,
+                certificate=provider_certificate.certificate,
+                ca=provider_certificate.ca,
+                chain=provider_certificate.chain,
+                revoked=revoked,
+                expiry_time=certificate_object.not_valid_after_utc,
+                expiry_notification_time=expiry_notification_time,
+            )
+            provider_certificates.append(provider_certificate)
+        return provider_certificates
+
+    def _request_certificate(self, csr: str, is_ca: bool) -> None:
+        """Add CSR to relation data."""
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            raise TLSCertificatesError(
+                f"Relation {self.relationship_name} does not exist - "
+                f"The certificate request can't be completed"
+            )
+        new_csr = CertificateSigningRequest(certificate_signing_request=csr.strip(), ca=is_ca)
+        app_or_unit = self._get_app_or_unit()
+        try:
+            requirer_relation_data = RequirerData.load(relation.data[app_or_unit])
+        except DataValidationError:
+            requirer_relation_data = RequirerData(
+                certificate_signing_requests=[],
+            )
+        new_relation_data = copy.deepcopy(requirer_relation_data.certificate_signing_requests)
+        new_relation_data.append(new_csr)
+        try:
+            RequirerData(certificate_signing_requests=new_relation_data).dump(
+                relation.data[app_or_unit]
+            )
+            logger.info("Certificate signing request added to relation data.")
+        except ModelError:
+            logger.warning("Failed to update relation data")
+
+    def _send_certificate_requests(self):
+        if not self.private_key:
+            logger.debug("Private key not generated yet.")
+            return
+        for certificate_request in self.certificate_requests:
+            if not certificate_request.is_valid():
```

Comment:
would it make sense to validate the CSR a bit earlier?
Maybe in the `init`? I'm not sure

Reply:
Of course it depends on what do we want to do:
- Prevent the requirer from adding `invalid` csrs
- Allow adding the invalid CSR but then not send it.

The first is a bit more aggressive so if we don't want to do it I understand it, and in that case please ignore this comment.

Reply:
Done